packages/service-utils/src/node/usageV2.ts

```
@@ -0,0 +1,110 @@
+
+  constructor(config: {
+    /**
+     * A descriptive name for your service. Example: "storage-server"
+     */
+    producerName: string;
+    /**
+     * The environment the service is running in.
+     */
+    environment: "development" | "production";
+
+    username: string;
+    password: string;
+  }) {
+    this.kafka = new Kafka({
+      clientId: `${config.producerName}-${config.environment}`,
+      brokers:
+        config.environment === "production"
+          ? ["warpstream.thirdweb.xyz:9092"]
+          : ["warpstream-dev.thirdweb.xyz:9092"],
+      ssl: {
+        checkServerIdentity(hostname, cert) {
+          return checkServerIdentity(hostname.toLowerCase(), cert);
+        },
+      },
+      sasl: {
+        mechanism: "plain",
+        username: config.username,
+        password: config.password,
+      },
+    });
+  }
+
+  /**
+   * Connect the producer.
+   * This must be called before calling `sendEvents()`.
+   */
+  async init() {
+    this.producer = this.kafka.producer({
+      allowAutoTopicCreation: false,
+    });
+    await this.producer.connect();
+  }
+
+  /**
+   * Send usageV2 events.
+   * This method may throw. To call this non-blocking:
+   *
+   * ```ts
+   * usageV2 = new UsageV2Producer(...)
+   * void usageV2.sendEvents(events).catch(console.error)
+   *
+   * @param events - The events to send.
+   */
+  async sendEvents(events: UsageV2Event[]): Promise<void> {
+    if (!this.producer) {
+      throw new Error("Producer not initialized. Call `init()` first.");
+    }
+
+    const parsedEvents = events.map(({ id, created_at, data, ...rest }) => {
+      return {
+        id: id ?? randomUUID(),
+        created_at: created_at ?? new Date(),
+        data: JSON.stringify(data),
+        ...rest,
+      };
+    });
+
+    await this.producer.send({
+      topic: TOPIC_USAGE_V2,
+      messages: parsedEvents.map((event) => ({
+        value: JSON.stringify(event),
+      })),
+    });
```

what should happen if this were to throw?

---

The caller needs to catch and handle throws. I don't want the library to silently ignore any exceptions. For example for AA we may want to retry usage emissions.